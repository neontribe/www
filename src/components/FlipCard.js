import React, { Children } from 'react'
import PropTypes from 'prop-types'
import css from 'styled-jsx/css'

const findChild = (children, { displayName }) =>
  Children.toArray(children).find(
    (child) => child.type.displayName === displayName
  )
const cloneWithProps = (node, props) => node && React.cloneElement(node, props)

const FlipCard = ({ children, isFlipped, animTime }) => (
  <div className="card-container">
    <div className="card-body">
      {cloneWithProps(findChild(children, CardBack), { isFlipped, animTime })}
      {/* Flip tell the front face to inform the animations */}
      {cloneWithProps(findChild(children, CardFront), { isFlipped, animTime })}
    </div>
    <style jsx>{`
      .card-body {
        width: 100%;
        height: 100%;
        display: flex;
      }

      .card-body::after {
        content: '';
        display: block;
        clear: both;
      }

      .card-container {
        height: 100%;
        transform-style: preserve-3d;
        position: relative;
        perspective: 1000px;
      }
    `}</style>
  </div>
)

FlipCard.propTypes = {
  // Whether the card is flipped or not
  isFlipped: PropTypes.bool,
  // The children, should be CardFront and CardBack
  children: PropTypes.node,
  // The time it should take for the animation to complete (in seconds)
  animTime: PropTypes.number,
}

FlipCard.defaultProps = {
  animTime: 0.7,
}

// Card sides, TODO: move to another file ---------------------
const cardSideStyles = css`
  .card-side {
    height: 100%;
    width: 100%;
    position: relative;
    backface-visibility: hidden;
    transform-style: preserve-3d;
  }

  .card-side + .card-side {
    margin-left: -100%;
  }
`

// Card front
const CardFront = ({ children, isFlipped, animTime }) => (
  <div className="card-side card-front">
    {children}
    <style jsx>{cardSideStyles}</style>
    <style jsx>{`
      .card-front {
        pointer-events: ${isFlipped ? 'none' : 'all'};
        transform: ${isFlipped ? 'rotateY(-180deg)' : 'rotateY(0deg)'}
          translateZ(0px);
        transition: all ${animTime}s linear;
      }
    `}</style>
  </div>
)
CardFront.displayName = 'FlipCardFront'
CardFront.propTypes = {
  // Whether the card is flipped or not
  isFlipped: PropTypes.bool,
  // The children, should be CardFront and CardBack
  children: PropTypes.node,
}

// Card back
const CardBack = ({ children, isFlipped, animTime }) => (
  <div className="card-side card-back" aria-hidden={!isFlipped}>
    {children}
    <style jsx>{cardSideStyles}</style>
    <style jsx>{`
      .card-back {
        transform: ${isFlipped ? 'rotateY(0deg)' : 'rotateY(180deg)'}
          translateZ(0px);
        transition: all ${animTime}s linear;
        pointer-events: ${isFlipped ? 'all' : 'none'};
      }
    `}</style>
  </div>
)
CardBack.displayName = 'FlipCardBack'
CardBack.propTypes = {
  // Whether the card is flipped or not
  isFlipped: PropTypes.bool,
  // The children, should be CardFront and CardBack
  children: PropTypes.node,
}

FlipCard.CardBack = CardBack
FlipCard.CardFront = CardFront

export { CardBack, CardFront }
export default FlipCard
