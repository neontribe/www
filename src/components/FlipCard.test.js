import React from 'react'
import { render } from '@testing-library/react'

import FlipCard, { CardFront, CardBack } from './FlipCard'

const TestFlipCard = () => (
  <FlipCard>
    <CardFront>front 1</CardFront>
    <CardBack>back 2</CardBack>
    Anything else
  </FlipCard>
)

test('displays the front and back', () => {
  const { getByText, queryByText } = render(<TestFlipCard />)

  expect(getByText('front 1')).toBeInTheDocument()
  expect(getByText('back 2')).toBeInTheDocument()
  expect(queryByText('Anything else')).not.toBeInTheDocument()
})

test('default render matches snapshot', () => {
  const { container } = render(<TestFlipCard />)

  expect(container).toMatchSnapshot()
})

test('isFlipped render matches snapshot', () => {
  const { container } = render(<TestFlipCard isFlipped />)

  expect(container).toMatchSnapshot()
})
