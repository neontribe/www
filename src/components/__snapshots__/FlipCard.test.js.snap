// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default render matches snapshot 1`] = `
<div>
  <div
    class="jsx-3761235168 card-container"
  >
    <div
      class="jsx-3761235168 card-body"
    >
      <div
        aria-hidden="true"
        class="jsx-3852085922 jsx-2114952886 card-side card-back"
      >
        back 2
      </div>
      <div
        class="jsx-3852085922 jsx-680621392 card-side card-front"
      >
        front 1
      </div>
    </div>
  </div>
</div>
`;

exports[`isFlipped render matches snapshot 1`] = `
<div>
  <div
    class="jsx-3761235168 card-container"
  >
    <div
      class="jsx-3761235168 card-body"
    >
      <div
        aria-hidden="true"
        class="jsx-3852085922 jsx-2114952886 card-side card-back"
      >
        back 2
      </div>
      <div
        class="jsx-3852085922 jsx-680621392 card-side card-front"
      >
        front 1
      </div>
    </div>
  </div>
</div>
`;
